packages/headless-react/src/client-utils.ts

```diff
@@ -1,23 +1,31 @@
 'use client';
 
-import {useEffect, useReducer, useRef} from 'react';
+import {useEffect, useReducer} from 'react';
+
+/**
+ * Subscriber is a function that takes a single argument, which is another function `listener` that returns `void`. The Subscriber function itself returns another function that can be used to unsubscribe the `listener`.
+ */
+export type Subscriber = (listener: () => void) => () => void;
+
+export type SnapshotGetter<T> = () => T;
 
 /**
  * Alternate for `useSyncExternalStore` which runs into infinite loops when hooks are used in `getSnapshot`
  * https://github.com/facebook/react/issues/24529
  */
 export function useSyncMemoizedStore<T>(
-  subscribe: (listener: () => void) => () => void,
-  getSnapshot: () => T
+  subscribe: Subscriber,
+  getSnapshot: SnapshotGetter<T>
 ) {
-  const stateRef = useRef(getSnapshot());
+  let latestSnapshot = getSnapshot();
+  const getLatestSnapshot = () => latestSnapshot;
```

Comment:
What was the bug that prompted these changes?

As far as I thought, we shouldn't use variables in a hook since they would be reinitialized on every render (in this case, calling getSnapshot every single time that anything in the parent component causes any re-render). I admittedly haven't tested that though.

When I originally tied to make the `useSyncMemoizedStore` hook, my implementation was significantly more complex (and probably not as efficient), maybe there's something missing from it that would solve your issue. Here it is:
<details>
<summary>use-sync-memoized-store.ts</summary>

```ts
import {DependencyList} from 'react';
import {useEffect, useReducer, useRef} from './wrapped';

function useHasChanged(deps: DependencyList) {
  const ref = useRef<DependencyList | null>(null);
  if (ref.current === null) {
    ref.current = deps;
    return false;
  }
  if (
    ref.current.length === deps.length &&
    !deps.some((dep, i) => !Object.is(ref.current![i], dep))
  ) {
    return false;
  }
  ref.current = deps;
  return true;
}

function useRefWithCallbackInitialState<T>(getInitialState: () => T) {
  const initialValueIsSet = useRef(false);
  const ref = useRef(null as T);
  !initialValueIsSet.current && (ref.current = getInitialState());
  initialValueIsSet.current = true;
  return ref;
}

/**
 * Same as [useSyncExternalStore](https://reactjs.org/docs/hooks-reference.html#usesyncexternalstore), but doesn't infinitely re-render with a memoized store.
 */
export function useSyncMemoizedStore<T>(
  subscribe: (listener: () => void) => () => void,
  getSnapshot: () => T
) {
  const state = useRefWithCallbackInitialState<T>(getSnapshot);
  const [, rerender] = useReducer((renderCount: number) => renderCount + 1, 0);
  useEffect(
    () => {
      let canRerender = false;
      const unsubscribe = subscribe(() => {
        if (!canRerender) {
          return;
        }
        state.current = getSnapshot();
        rerender();
      });
      canRerender = true;
      return unsubscribe;
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [subscribe]
  );
  if (useHasChanged([getSnapshot])) {
    state.current = getSnapshot();
  }
  return state.current as T;
}
```
</details>

Reply:
Without the fixes in 48d0af43f06573733eacd405e0f32d790b4ea338 the tests fail as follows
```
 FAIL  src/client-utils.test.tsx
  ● useSyncMemoizedStore › should update the state when the store changes

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 1

      Object {
    -   "count": 1,
    +   "count": 0,
      }

      32 |     expect(result.current).toEqual(snapshot1);
      33 |     rerender();
    > 34 |     expect(result.current).toEqual(snapshot2);
         |                            ^
      35 |   });
      36 |
      37 |   test('should unsubscribe and re-subscribe to new function when subscribe function is changed', () => {

      at Object.<anonymous> (src/client-utils.test.tsx:34:28)

  ● useSyncMemoizedStore › should replace current snapshot when getSnapshot function is changed

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 1

      Object {
    -   "count": 1,
    +   "count": 0,
      }

      72 |     expect(result.current).toEqual(snapshot1);
      73 |     rerender({getSnapshot: getSnapshot2});
    > 74 |     expect(result.current).toEqual(snapshot2);
         |                            ^
      75 |   });
      76 |
      77 |   it('should re-render only when the subscribe listener is called', () => {

      at Object.<anonymous> (src/client-utils.test.tsx:74:28)

 PASS  src/search-engine.test.tsx (5.087 s)

Test Suites: 1 failed, 1 passed, 2 total
Tests:       2 failed, 14 passed, 16 total
```

With the code you have provided 1 test fails
```
 FAIL  src/client-utils.test.tsx
  ● useSyncMemoizedStore › should update the state when the store changes

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 1

      Object {
    -   "count": 1,
    +   "count": 0,
      }

      32 |     expect(result.current).toEqual(snapshot1);
      33 |     rerender();
    > 34 |     expect(result.current).toEqual(snapshot2);
         |                            ^
      35 |   });
      36 |
      37 |   test('should unsubscribe and re-subscribe to new function when subscribe function is changed', () => {

      at Object.<anonymous> (src/client-utils.test.tsx:34:28)

 PASS  src/search-engine.test.tsx (5.009 s)

Test Suites: 1 failed, 1 passed, 2 total
Tests:       1 failed, 15 passed, 16 total
```

As the current code passes both the unit and e2e tests and is simpler, tempted to leave it as is for now.

Reply:
I'm hesitant, since in our case, `getSnapshot` calls a getter (someController.state). For some controllers, this getter is fairly simple, but either way it'll always generate an object.

I tested, and was able to confirm that calling the method outside of a hook like `useMemo` causes it to be called on every re-render of the parent component (see https://7wvpvj-5173.csb.app/, which logs "Called getSnapshot1" whenever the state is updated every 5 milliseconds, but doesn't log "Called getSnapshot2" repeatedly since it's wrapped in a useMemo hook).

Concerning the first test failure ("should update the state when the store changes"), I suspect it may be failing due to how the test is setup, specifically this line:
```ts
const getSnapshot = jest
  .fn()
  .mockReturnValueOnce(snapshot1)
  .mockReturnValueOnce(snapshot2);
```

As far as I know, the reference to `getSnapshot` never changes, only its returned value, which IMO we don't need to verify (and shouldn't). Instead, I think this test should verify whether replacing the `getSnapshot` function with a different one updates the state.

Reply:
Thanks for testing out with an example. But in this case the `getStaticState` from `useCallback` is passed into `useSyncMemoizedStore` - so maybe the memoized return value is used instead?
I have tried out `useRef`, `useMemo` and even `useState` to wrap `getSnapshot()` - but all of them fail tests as indicated above. Is there a specific reason you believe the code is problematic when the tests are passing? Is there any issue with the tests?

Reply:
you were right! added test to test getSnapshot is not called on rerender, updated to use useMemo, fixed another test that was not testing the right thing.